### ISSUE TITLE
problem with .ckpt file 

### ISSUE THREAD

**Issue by gilad011** [GeeeG/CloudPose#12]
Hi,

1. what is the difference between:
general.add_argument(**'--num_point**', type=int, default=1024, help='Point Number [256/512/1024] [default: 256]')
general.add_argument('**--total_num_point**', type=int, default=1024, help='Dataset Point Number [256/512/1024] [default: 1024]')?
2. How can I evaluate if the loss values i get during training are good or bad?
for example: trans_los of 100 and axag_los of 0.99, what is the meaning of these values?

**Comment by GeeeG** [comment]
1. total_num_point is the number of point for each point cloud segment saved in the tfrecord file, you need to know this to load the dataset properly
num_point is the one that you actually use for training and testing
num_point <= total_num_point

2. the unit for trans loss is meter, so 100 is 100 meters, the unit for axag_loss is radian, so 0.1 radian is around 5.7 degrees

**Comment by gilad011** [comment]
Thanks!

Another few questions:

1. During training, I can see that the trans_loss values are starting to decay, but the axag_loss remain steady. Have you came across this phenomena? 
How would you approach to solve it? I would like to encounter that my dataset is in milimeters units, could it explain the phenomena?

2. I saw that the total loss is a weighted combination between the trans_loss and the axag_loss. In fact:

Total_loss = 10*trans_loss + axag_loss.

Why you chose 10? Is it rely on some theory? 

Thanks again!

**Comment by GeeeG** [comment]
1. Two possible reasons. First, the learning for rotation takes time to converge, so if you leave it training for longer, the loss might start to drop. Second, the network can't learn rotation for some reason. What kind of object are you using? The milimeter units shouldn't be a reason, but can you scale it to meter?
2. 10 is given by the ratio between the expected error of translation and rotation at the end of the training

**Comment by gilad011** [comment]
Hi,

The object that im using is looking like a rectangle plastic box cover with grooves in it.

Im training the network solely on this object.

Currently, im training according to 2500 samples per this object.

Still, cannot make the network learn rotations. I got axag_loss of arround 2.2 radians even after 500 epoches. While the translation loss converges nicely.

Why is it happenning? Do you think that i need to change architecture? 
Anyway, i should be able to get overfit on the training data, doesnt it?

**Comment by GeeeG** [comment]
Hi, one problem might be the rotational symmetric shape of the box. Have you visually inspect the results? For rotational symmetric objects, sometime the axag loss is large, but the ADDS score is low.
I would suggest first do a visual inspection, projecting your object model into the camera coordinate with the estimated poses (Rx_oject + t = x_camera), to see how good/bad it overlays. If it's not nicely overlayed, I would suggest to train with ADDS score instead of (axag + translation). I can provide you with the code for this.

**Comment by gilad011** [comment]
Thanks! What's the definition for ADDS in our case?
I Would like to try your suggestion, could you provide the code? 

Is it possible to add to the loss the distance from the CAD model point cloud? Have you tried it?

**Comment by GeeeG** [comment]
For the ADD and ADDs definition, you can take a look at this paper: On Evaluation of 6D Object Pose Estimation, http://cmp.felk.cvut.cz/~hodanto2/data/hodan2016evaluation.pdf
They are both matric for 6D pose, including both rotation and translation.
Here's the tf implemntation from me, the model_gt is you take the object model and apply ground truth 6D pose to it, and the model_pred is transformed with prediction pose, I planned to also update this in git with a complete examples but haven't found the time it:

def transform_object_model(model, rot_mat, trans):
    model_rot = tf.matmul(rot_mat, tf.transpose(model, perm=[0, 2, 1]))
    model_rot = tf.transpose(model_rot, perm=[0, 2, 1])
    model_transformed = model_rot + tf.expand_dims(trans, axis=1)
    return model_transformed

def ADD_loss(model_gt, model_pred):
    distances = tf.norm(model_gt-model_pred, axis=2, ord='euclidean')
    return tf.reduce_mean(distances, axis=1)

def ADS_loss(model_gt, model_pred):
    I2, I1 = tf.meshgrid(tf.range(model_gt.shape[1]), tf.range(model_pred.shape[1]))
    I1 = tf.reshape(I1, [-1])
    I2 = tf.reshape(I2, [-1])
    points_gt = tf.gather(model_gt, I1, axis=1)
    points_pred = tf.gather(model_pred, I2, axis=1) # B,32*32,3
    distances = tf.norm(points_gt - points_pred, axis=2, ord='euclidean') # B,32, 32
    distances = tf.reshape(distances, [-1, model_gt.shape[1], model_pred.shape[1]]) # B, 32, 32
    distances = tf.reduce_min(distances, axis=2)
    return tf.reduce_mean(distances, axis=1)

**Comment by Windson9** [comment]
Hi @gilad011. Were you able to train the network properly? I am facing the same issue where the axag loss is not converging.

**Comment by Windson9** [comment]
Hi @GeeeG. I have trained the model using the TF records provided in this repo. The loss graphs are optimal and the visual inspection while evaluation is also okay. However when I run the inference with my custom test sample. The object model is not overlaying properly with the segmented one. I am using the same model which is giving great results on test sample provided in this repo.

**Comment by Windson9** [comment]
Hi @GeeeG. I have trained the model using the TF records provided in this repo. The loss graphs are optimal and the visual inspection while evaluation is also okay. However when I run the inference with my custom test sample. The object model is not overlaying properly with the segmented one. I am using the same model which is giving great results on test sample provided in this repo.
